File: libs/python/mail_action_usps/parse_digest.py
#!/usr/bin/env python3
"""Parse USPS Informed Delivery digest HTML to extract structured mailpiece data."""

import json
import os
import re
from pathlib import Path
from html.parser import HTMLParser


class MailpieceExtractor(HTMLParser):
    """Extract mailpiece info from USPS digest HTML body."""
    
    def __init__(self):
        super().__init__()
        self.mailpieces = []
        self.packages = []
        self._in_from = False
        self._current_from = ""
        self._capture_text = False
        self._text_buf = ""
        self._mail_count = None
        self._pkg_count = None
    
    def handle_data(self, data):
        text = data.strip()
        if not text:
            return
        
        # Look for mail count
        m = re.match(r"You have (\d+) mailpiece", text)
        if m:
            self._mail_count = int(m.group(1))
        
        # Look for package count
        m = re.match(r"You have (\d+) package", text)
        if m:
            self._pkg_count = int(m.group(1))
        
        # Capture FROM labels
        if text.startswith("FROM:"):
            self._current_from = text[5:].strip()
        elif self._in_from:
            self._current_from += " " + text


def parse_digest_html(html_path: str) -> dict:
    """
    Parse a USPS digest HTML file and extract:
    - Mail count
    - Package count  
    - FROM labels for each mailpiece
    - Image CID references
    - Tracking numbers
    """
    with open(html_path, "r", errors="replace") as f:
        html = f.read()
    
    result = {
        "mail_count": 0,
        "package_count": 0,
        "from_labels": [],
        "image_cids": [],
        "tracking_numbers": [],
        "has_no_image": False,
    }
    
    # Extract mail count
    m = re.search(r"You have (\d+) mailpiece", html)
    if m:
        result["mail_count"] = int(m.group(1))
    
    # Extract package count
    m = re.search(r"You have (\d+) package", html)
    if m:
        result["package_count"] = int(m.group(1))
    
    # Extract FROM labels
    from_matches = re.findall(r'(?:FROM|From):\s*([^<\n]+)', html)
    result["from_labels"] = [f.strip() for f in from_matches if f.strip()]
    
    # Extract CID image references (these are the inline mailpiece scans)
    cid_matches = re.findall(r'src="cid:([^"]+)"', html)
    result["image_cids"] = cid_matches
    
    # Extract tracking numbers (USPS format: 20-34 digits)
    tracking_matches = re.findall(r'\b((?:94|92|93|94)\d{18,30})\b', html)
    result["tracking_numbers"] = list(set(tracking_matches))
    
    # Check for "no image" placeholder text
    if "A scanned image of this mail piece" in html and "not available" in html:
        result["has_no_image"] = True
    
    return result


def parse_all_digests(base_dir: str) -> dict:
    """Parse all digest folders and return structured data keyed by date."""
    base = Path(base_dir)
    all_data = {}
    
    for date_dir in sorted(base.iterdir()):
        if not date_dir.is_dir() or not date_dir.name.startswith("20"):
            continue
        
        body_path = date_dir / "body.html"
        if not body_path.exists():
            continue
        
        parsed = parse_digest_html(str(body_path))
        
        # List actual image files
        images = sorted([
            f.name for f in date_dir.iterdir()
            if f.suffix == ".jpg" and not f.name.startswith("content-")
        ])
        
        parsed["date"] = date_dir.name
        parsed["images"] = images
        parsed["scan_images"] = [i for i in images if re.match(r'\d{10}-\d{3}\.jpg', i)]
        parsed["ad_images"] = [i for i in images if i.startswith("mailer-")]
        
        all_data[date_dir.name] = parsed
    
    return all_data


if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description="Parse USPS digest HTML files")
    parser.add_argument("input_dir", help="Directory containing date-based digest folders")
    parser.add_argument("--json", "-j", action="store_true", help="Output as JSON")
    args = parser.parse_args()
    
    data = parse_all_digests(args.input_dir)
    
    if args.json:
        print(json.dumps(data, indent=2))
    else:
        for date_str, info in sorted(data.items()):
            print(f"\n{date_str}:")
            print(f"  Mail pieces: {info['mail_count']}")
            print(f"  Packages: {info['package_count']}")
            print(f"  Scan images: {len(info['scan_images'])}")
            print(f"  Ad images: {len(info['ad_images'])}")
            if info['from_labels']:
                print(f"  FROM labels: {', '.join(info['from_labels'][:5])}")
            if info['tracking_numbers']:
                print(f"  Tracking: {', '.join(info['tracking_numbers'])}")

File: libs/ts/mail_action_usps/src/parse-digest.ts
/**
 * Parse USPS Informed Delivery digest HTML to extract structured mailpiece data.
 */

import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, extname, basename } from "node:path";

export interface DigestResult {
  mail_count: number;
  package_count: number;
  from_labels: string[];
  image_cids: string[];
  tracking_numbers: string[];
  has_no_image: boolean;
}

export interface ParsedDigest extends DigestResult {
  date: string;
  images: string[];
  scan_images: string[];
  ad_images: string[];
}

/**
 * HTML parser that extracts mailpiece info from USPS digest HTML body.
 * Ports the Python MailpieceExtractor (html.parser.HTMLParser subclass).
 */
export class MailpieceExtractor {
  mailpieces: Record<string, unknown>[] = [];
  packages: Record<string, unknown>[] = [];
  private _currentFrom = "";
  private _mailCount: number | null = null;
  private _pkgCount: number | null = null;

  parse(html: string): void {
    // Walk through text content between tags
    let pos = 0;
    while (pos < html.length) {
      const tagStart = html.indexOf("<", pos);
      if (tagStart === -1) {
        this._handleData(html.slice(pos));
        break;
      }
      if (tagStart > pos) {
        this._handleData(html.slice(pos, tagStart));
      }
      const tagEnd = html.indexOf(">", tagStart);
      if (tagEnd === -1) break;
      pos = tagEnd + 1;
    }
  }

  private _handleData(data: string): void {
    const text = data.trim();
    if (!text) return;

    const mailMatch = text.match(/You have (\d+) mailpiece/);
    if (mailMatch) {
      this._mailCount = parseInt(mailMatch[1], 10);
    }

    const pkgMatch = text.match(/You have (\d+) package/);
    if (pkgMatch) {
      this._pkgCount = parseInt(pkgMatch[1], 10);
    }

    if (text.startsWith("FROM:")) {
      this._currentFrom = text.slice(5).trim();
    }
  }

  get mailCount(): number | null {
    return this._mailCount;
  }

  get pkgCount(): number | null {
    return this._pkgCount;
  }
}

/**
 * Parse a USPS digest HTML file and extract mail/package counts,
 * FROM labels, image CID references, and tracking numbers.
 */
export function parseDigestHtml(htmlPath: string): DigestResult {
  const html = readFileSync(htmlPath, { encoding: "utf-8" });

  const result: DigestResult = {
    mail_count: 0,
    package_count: 0,
    from_labels: [],
    image_cids: [],
    tracking_numbers: [],
    has_no_image: false,
  };

  // Extract mail count
  const mailMatch = html.match(/You have (\d+) mailpiece/);
  if (mailMatch) {
    result.mail_count = parseInt(mailMatch[1], 10);
  }

  // Extract package count
  const pkgMatch = html.match(/You have (\d+) package/);
  if (pkgMatch) {
    result.package_count = parseInt(pkgMatch[1], 10);
  }

  // Extract FROM labels
  const fromMatches = html.matchAll(/(?:FROM|From):\s*([^<\n]+)/g);
  for (const m of fromMatches) {
    const label = m[1].trim();
    if (label) {
      result.from_labels.push(label);
    }
  }

  // Extract CID image references (inline mailpiece scans)
  const cidMatches = html.matchAll(/src="cid:([^"]+)"/g);
  for (const m of cidMatches) {
    result.image_cids.push(m[1]);
  }

  // Extract tracking numbers (USPS format: 20-34 digits)
  const trackingMatches = html.matchAll(/\b((?:94|92|93|94)\d{18,30})\b/g);
  const trackingSet = new Set<string>();
  for (const m of trackingMatches) {
    trackingSet.add(m[1]);
  }
  result.tracking_numbers = [...trackingSet];

  // Check for "no image" placeholder text
  if (
    html.includes("A scanned image of this mail piece") &&
    html.includes("not available")
  ) {
    result.has_no_image = true;
  }

  return result;
}

/**
 * Parse all digest folders and return structured data keyed by date.
 */
export function parseAllDigests(
  baseDir: string,
): Record<string, ParsedDigest> {
  const allData: Record<string, ParsedDigest> = {};

  let entries: string[];
  try {
    entries = readdirSync(baseDir).sort();
  } catch {
    return allData;
  }

  for (const name of entries) {
    if (!name.startsWith("20")) continue;
    const dateDirPath = join(baseDir, name);

    try {
      if (!statSync(dateDirPath).isDirectory()) continue;
    } catch {
      continue;
    }

    const bodyPath = join(dateDirPath, "body.html");
    try {
      statSync(bodyPath);
    } catch {
      continue;
    }

    const parsed = parseDigestHtml(bodyPath);

    // List actual image files
    const dirFiles = readdirSync(dateDirPath).sort();
    const images = dirFiles.filter(
      (f: string) => extname(f) === ".jpg" && !f.startsWith("content-"),
    );
    const scanImages = images.filter((i: string) => /^\d{10}-\d{3}\.jpg$/.test(i));
    const adImages = images.filter((i: string) => i.startsWith("mailer-"));

    allData[name] = {
      ...parsed,
      date: name,
      images,
      scan_images: scanImages,
      ad_images: adImages,
    };
  }

  return allData;
}
